<img src=cenapad.jpg align=left><BR><BR><BR><BR><BR><BR>
<h1>Curso Introducción al Lenguaje Python<br>Programación Estructurada</h1>
<img src=python-logo.png align=left>

# 2.1 Primeros pasos hacia la programación¶

Por supuesto, podemos usar Python para tareas más complicadas que sumar dos y dos. Por ejemplo, podemos escribir una subsecuencia inicial de la serie de Fibonacci así:

In [ ]:
# Series de Fibonacci:
# la suma de dos elementos define el siguiente
a, b = 0, 1
while b < 10:
    print(b)
    a, b = b, a+b

In [ ]:
lista=[]
a, b = 0, 1

while b < 10:
    lista.append(b)
    a, b = b, a+b
    
print(lista)

Este ejemplo introduce varias características nuevas.
<ul>
<li>La primer línea contiene una asignación múltiple: las variables a y b toman en forma simultanea los nuevos valores 0 y 1. En la última linea esto se vuelve a usar, demostrando que las expresiones a la derecha son evaluadas antes de que suceda cualquier asignación. Las expresiones a la derecha son evaluadas de izquierda a derecha.</li><br>

<li>El bucle while se ejecuta mientras la condición (aquí: b < 10) sea verdadera. En Python, como en C, cualquier entero distinto de cero es verdadero; cero es falso. La condición también puede ser una cadena de texto o una lista, de hecho cualquier secuencia; cualquier cosa con longitud distinta de cero es verdadero, las secuencias vacías son falsas. La prueba usada en el ejemplo es una comparación simple.</li><br>

<li>El cuerpo del bucle tiene sangría: Python utiliza 4 espacios para diferencias que se encuentra en un bucle o función. El interpretador permite la utilziación de 4 espacios o tabulación (tab) pero no mezclar ambos tipos, pues puede generar errores en la interpretación.</li><br>

<li>La función print() escribe el valor de el o los argumentos que se le pasan. Difiere de simplemente escribir la expresión que se quiere mostrar en la forma en que maneja múltiples argumentos, cantidades en punto flotante, y cadenas. Las cadenas de texto son impresas sin comillas, y un espacio en blanco es insertado entre los elementos, así puedes formatear cosas de una forma agradable:</li>

</ul>

In [ ]:
i = 256 * 256
print('El valor de i es', i)

El parámetro nombrado end puede usarse para evitar el salto de linea al final de la salida, o terminar la salida con una cadena diferente:

In [ ]:
lista = []
a, b = 0, 1
while b < 1000:
    print(b, end=', ')
    lista.append(b)
    a, b = b, a+b

In [ ]:
for elemento in lista:
    print(elemento, end=", ")

Otros ejemplos de la utilización de <b>print</b>.

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ')
print('Mars', 'Jupiter', 'Saturn', sep=', ')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ', end='\t\t')
print('Mars', 'Jupiter', 'Saturn', sep=', ', end='\t\t')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ')
print('Mars', 'Jupiter', 'Saturn', sep=', ')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Printing in a Nutshell', end='\n * ')
print('Calling Print', end='\n * ')
print('Separating Multiple Arguments', end='\n * ')
print('Preventing Line Breaks')

Para más información sobre print visitar: https://realpython.com/python-print/

# 2.2. Condicionales: if, elif y else

La declaración if/elif/else funciona para controlar el fluxo de ejecución basandose en una o más condiciones. La sintaxis es la siguiente:

In [ ]:
#No ejecute está célula
if condicion:
    comando1
    comando2
    ...
    comandoN
elif condicion:
    comando1
    comando2
    ...
    comandoN
else:
    comando1
    comando2
    ...
    comandoN

Observa la <b>IDENTACIÓN</b> que define los bloques de código, a diferencia de otros lenguajes que usan delimitadores como llaves tales como "{" y "}" o bloques con "begin" y "end".<br>

Una buena práctica es usar <b>IDENTATION</b> con 4 espacios (primer nivel), 8 espacios (segundo nivel), etc.<br>

<b>NO SE DEBÉN USAR TABULACIONES</b> (debido a las convenciones de portabilidad y legibilidad), aunque están permitidas.<br>

<b>NUNCA se debe mezclar espacios con tabulaciones en el mismo programa.</b><br>

IDENTATION en Python es obligatorio y forma parte de la sintaxis. La ventaja de esta característica es la legibilidad del código. La mayoría de los editores de código o IDE de desarrollo facilitarán el trabajo de IDENTACIÓN colocando espacios automáticamente si se encuentra dentro del contexto de un bloque de código.

#### Ejemplo de uso del condicional if elif else:

In [ ]:
nombre = input("¿Cual es su nombre?:")
edad = int(input("¿Cual es su edad?:")) #Ingrese un número entero
grupo=""

if edad <= 12:
    grupo = "Infancia"
elif edad <= 18:
    grupo = "Adolescencia"
elif edad <= 26:
    grupo = "Juventud"
elif edad <= 65:
    grupo = "Adultez"
else:
    grupo = "Persona Mayor"

print(f"Nombre: {nombre} - Edad: {edad} - Pertenece al grupo {grupo}.")

La palabra reservada ‘elif‘ es una abreviación de ‘else if’, y es útil para evitar un sangrado excesivo. Una secuencia if ... elif ... elif ... sustituye las sentencias switch o case encontradas en otros lenguajes.

La declaración if/elif/else puede precisa tener  tener ninguún bloque elif, por ejemplo:

In [ ]:
edad=21
if edad<=17:
    print("Menor de edad")
else:
    print("Mayor de edad")

Es posible combinar condiones con otros operadores <b>and</b>, <b>not</b> y <b>or</b>.

In [ ]:
edad = 31
if edad <= 12:
    print("Infancia")
elif edad > 12 and edad < 21:
    print("Adolescente")
elif edad>=21:
    print("Adulto")

In [ ]:
x = int(input("Ingrese un número entero: "))
y = int(input("Ingrese otro número entero: "))
if x%2==1 and y%2==1:
    print("Ambos números son impares")
else:
    print("Por lo menos uno de los números es par")

In [ ]:
x = int(input("Ingrese un número entero: "))
y = int(input("Ingrese otro número entero: "))
if x%2==1 or y%2==1:
    print("Por lo menos uno delos números es impar")
else:
    print("Ambos números son pares")

In [ ]:
num = list(range(1,30))
for n in num:
    if n%3==0:
        print(f"{n} es divisible por 3")

Alternativamente, en lugar de imprimir en la pantalla, podemos crear una lista con los números divisibles por 3 y luego imprimirla

In [ ]:
num = list(range(1,30))
num_div3=[] #Iniciando una lista vacia (Esto es necesario para que podemos usar el método append)
for n in num:
    if n%3==0:
        num_div3.append(n)
print(num_div3)

### Ejercicio 1:

Dado tres puntos en el plano expresados como coordenadas (x, y) informar cuál es el que se encuentra más lejos del centro de coordenadas.

In [12]:
import math 
#math.sqrt
coordenadas1 = input("Introduce la coordenada 1 en formato (x,y): ")
coordenadas2 = input("Introduce la coordenada 2 en formato (x,y): ")
coordenadas3 = input("Introduce la coordenada 3 en formato (x,y): ")

coordenadas1 = coordenadas1.replace("(" , "")
coordenadas1 = coordenadas1.replace(")" , "")
coordenadas2 = coordenadas2.replace("(" , "")
coordenadas2 = coordenadas2.replace(")" , "")
coordenadas3 = coordenadas3.replace("(" , "")
coordenadas3 = coordenadas3.replace(")" , "")

lista1 = coordenadas1.split(',')
lista2 = coordenadas1.split(',')
lista3 = coordenadas1.split(',')
valor1 = math.sqrt(int(lista1[0])* int(lista1[0]) + (int(lista1[1]) * int(lista1[1])))
valor2 = math.sqrt(int(lista2[0])* int(lista2[0]) + (int(lista2[1]) * int(lista2[1])))
valor3 = math.sqrt(int(lista3[0])* int(lista3[0]) + (int(lista3[1]) * int(lista3[1])))
if(valor1>valor2):
    if(valor1>valor3):
        print("La cordenada 1 se encuentra mas lejos")
    else:
        print("La cordenada 3 se encuentra mas lejos")
else:
    if(valor2>valor3):
        print("La cordenada 2 se encuentra mas lejos")
    else:
        print("La cordenada 3 se encuentra mas lejos")
    

Introduce la coordenada 1 en formato (x,y): (3,2)
Introduce la coordenada 2 en formato (x,y): (1,5)
Introduce la coordenada 3 en formato (x,y): (4,2)
La cordenada 3 se encuentra mas lejos


### Ejercicio 2:

Escribir un programa que almacene la cadena de caracteres contraseña en una variable, pregunte al usuario por la contraseña e imprima por pantalla si la contraseña introducida por el usuario coincide con la guardada en la variable sin tener en cuenta mayúsculas y minúsculas.

Sugerencia: utilizar la función lower.

<b>Salida deseada:</b><br>
La contaseña coincide<br>
<br>
La contraseña no coincide

In [3]:
contraseña = "hola"
contraseñaInput = input("Introduce la contraseña: ")

if(contraseñaInput.lower() == contraseña):
    print("La contaseña coincide")
else:
    print("La contraseña no coincide")


Introduce la contraseña: HOla
La contaseña coincide


### Ejercicio 3:

Para tributar un determinado impuesto se debe ser mayor de 16 años y tener unos ingresos iguales o superiores a 2000 Bs mensuales. Escribir un programa que pregunte al usuario su edad y sus ingresos mensuales y muestre por pantalla si el usuario tiene que tributar o no.

<b>Salida deseada:</b><br>
Tienes que tributar<br>
<br>
No tienes que tributar

In [14]:

edad = int(input("edad: "))
ingresos = float(input("Ingresos: "))

if(edad > 16 and ingresos>=2000 ):
    print("Tienes que tributar")
else:
    print("No tienes que tributar")



edad: 16
Ingresos: 3000
No tienes que tributar


### Ejercicio 4:

Escriba un programa que pida dos números enteros y que calcule su división, escribiendo si la división es exacta o no. El programa debe tomar en cuenta que no se puede dividir por cero.

In [18]:
x1 = int(input("numero entero 1: "))
x2 = int(input("numero entero 2: "))
if(x2 == 0 ):
    print('infinito')
else:
    if(x1%x2 == 0):
        print('Es division exacta')
    else:
        print('No es division exacta')


numero entero 1: 1
numero entero 2: 0
infinito


### Ejercicio 5:

Escriba un programa que pida año, mes y día y retorne cuantos días han pasado o faltan a la fecha actual.

In [21]:
a = int(input("Introduce año "))
m = int(input("Introduce mes "))
d = int(input("Introduce dia "))
dias = (a - 2020)*365
dias += (m -2)*30
dias += d - 27
if(a > 0 and m>0 and d>0):
    if(dias>= 0 ):
        print('Pasaron ', dias, ' de la fecha actual')
    else:
        print('Faltan ', dias*-1, ' para la fecha actual')
else:
    print('No existe el año,mes,dia 0 o negativo.')

Introduce año 2012
Introduce mes 2
Introduce dia 1
Faltan  2946  para la fecha actual


### Ejercicio 6:

Escriba un programa que pida dos números enteros y que escriba si el mayor es múltiplo del menor. El programa debe avisar cuando se escriba valores negativos o nulos.

In [25]:
num1 = int(input("Entero 1 "))
num2 = int(input("Entero 2 "))
if(num1>0 or num2>0):
    if(num1> num2 ):
        if(num1%num2 ==0):
            print('El numero ', num1 , ' es multiplo del numero ' , num2)
    else:
        if(num2%num1 ==0):
            print('El numero ', num2 , ' es multiplo del numero ' , num1)
else:
    print('deben ser positivos o no nulos.')

Entero 1 10
Entero 2 20
El numero  20  es multiplo del numero  10


### Ejercicio 7:

Escriba un programa que pida los coeficientes de una ecuación de segundo grado (a x² + b x + c = 0) y escriba la solución.

Se recuerda que una ecuación de segundo grado puede no tener solución (solución con números complejos), tener una solución única, tener dos soluciones o que todos los números sean solución. Se recuerda que la fórmula de las soluciones cuando hay dos soluciones es x = (-b ± √(b2-4ac) ) / (2a)

En caso de no tener solución, hacer uso de la números complejos de Python.

Más información en:<br>
[1] https://relopezbriega.github.io/blog/2015/10/12/numeros-complejos-con-python/<br>
[2] https://pybonacci.org/2014/02/04/como-separar-parte-real-e-imaginaria-en-sympy/

<b>Salida deseada</b><br>
ECUACIÓN A X² + B X + C = 0<br>
Escriba el valor del coeficiente a: 2<br>
Escriba el valor del coeficiente b: -7<br>
Escriba el valor del coeficiente c: 3<br>
La ecuación tiene dos soluciones: 3.0 y 0.5

In [1]:
import math
print('Salida deseada')
print("ECUACIÓN A X² + B X + C = 0")
a = float(input('Escriba el valor del coeficiente a: '))
b = float(input('Escriba el valor del coeficiente b: '))
c = float(input('Escriba el valor del coeficiente c: '))

if( b*b-4*a*c >=0):
    x1 = (-b + math.sqrt(b*b-4*a*c) ) / (2*a)
    x2 = (-b - math.sqrt(b*b-4*a*c) ) / (2*a)
    if(x1==x2):
        print("La ecuación tiene la solucion " , x1 )
    else:
        print("La ecuación tiene dos soluciones " , x1 , " y ", x2)
else:
    ima = math.sqrt( (b*b-4*a*c)*-1 )/(2*a)  
    x1 = (-b) / (2*a)
    print("La ecuación tiene solucion compleja " , x1 , " y ", ima , 'j')


Salida deseada
ECUACIÓN A X² + B X + C = 0
Escriba el valor del coeficiente a: 1
Escriba el valor del coeficiente b: 1
Escriba el valor del coeficiente c: 1
La ecuación tiene solucion compleja  -0.5  y  0.8660254037844386 j


### Ejercicio 8:

Escriba un programa que pregunte primero si se quiere calcular el área de un triángulo o la de un círculo. Si se contesta que se quiere calcular el área de un triángulo (escribiendo T o t), el programa tiene que pedir entonces la base y la altura y escribir el área. Si se contesta que se quiere calcular el área de un círculo (escribiendo C o c), el programa tiene que pedir entonces el radio y escribir el área.

Se recuerda que el área de un triángulo es base por altura dividido por 2 y que el área de un círculo es Pi (aproximadamente 3,141592) por el radio al cuadrado. Utilizar 2 decimales para la salida.

Nota: Utilice la libreria math para obtener la función pi.

<b>Salida deseada:</b><br>
CÁLCULO DE ÁREAS<br>
Elija una figura geométrica:<br>
a) Triángulo<br>
b) Círculo<br>
¿Qué figura quiere calcular (Escriba T o C)? T<br>
Escriba la base: 3<br>
Escriba la altura: 5.5<br>
Un triángulo de base 3.0 y altura 5.0 tiene un área de 8.25<br>
<br>
CÁLCULO DE ÁREAS<br>
Elija una figura geométrica:<br>
a) Triángulo<br>
b) Círculo<br>
¿Qué figura quiere calcular (Escriba T o C)? C<br>
Escriba el radio: 2<br>
Un círculo de radio 2.0 tiene un área de 12.57<br>

In [7]:
import math
print('CÁLCULO DE ÁREAS')
print('Elija una figura geométrica:')
print('a) Triángulo')
print('b) Círculo')
figura = input('¿Qué figura quiere calcular (Escriba T o C)? ')
if(figura.lower() == 't'):
    base = float(input('Escriba la base: '))
    altura = float(input('Escriba la altura: '))
    print('un triangulo de base ',base,' y altura ',altura, ' tiene un area de ',"{0:.2f}".format((base*altura)/2 ))
if(figura.lower() == 'c'):
    radio = float(input('Escriba el radio: '))
    print('Un circulo de radio ',radio,' tiene un area de ',"{0:.2f}".format(math.pi*radio*radio) )
if(figura.lower() != 'c' and figura.lower() != 't'):
    print('No existe esa opcion')
    
    

CÁLCULO DE ÁREAS
Elija una figura geométrica:
a) Triángulo
b) Círculo
¿Qué figura quiere calcular (Escriba T o C)? t
Escriba la base: 4
Escriba la altura: 2.1
un triangulo de base  4.0  y altura  2.1  tiene un area de  4.20


# 2.3. La Sentencia for

La sentencia <b>for</b> en Python difiere un poco de lo que uno puede estar acostumbrado en lenguajes como C o Pascal. En lugar de siempre iterar sobre una progresión aritmética de números (como en Pascal) o darle al usuario la posibilidad de definir tanto el paso de la iteración como la condición de fin (como en C), la sentencia <b>for</b> de Python itera sobre los ítems de cualquier secuencia (una lista o una cadena de texto), en el orden que aparecen en la secuencia. Por ejemplo:

In [2]:
# Midiendo cadenas de texto
palabras = ['gato', 'ventana', 'defenestrado']
for p in palabras:
    print(p, len(p))

gato 4
ventana 7
defenestrado 12


Si necesitá modificar la secuencia sobre la que estás iterando mientras estás adentro del ciclo (por ejemplo para borrar algunos ítems), se recomienda que hagas primero una copia. Iterar sobre una secuencia no hace implícitamente una copia. La notación abajo es especialmente conveniente para esto:

In [3]:
for p in palabras[:]:  # hace una copiade toda la lista
    if len(p) > 6:
        palabras.insert(0, p)

print(palabras)

['defenestrado', 'ventana', 'gato', 'ventana', 'defenestrado']


Con for w in words:, el ejemplo intentaría crear una lista infinita, insertando defenestrado una y otra vez.<br>

Un bucle <b>for</b> funciona como un iterador. Un <b>for</b> se utiliza para realizar acciones sobre elementos de una secuencia o cualquier elemento iterable, como cadenas, listas y otros objetos iterables que veremos más adelante.

<b>Formato general para una declaración for</b>

for item in object:<br>
&nbsp;&nbsp;&nbsp;&nbsp;statements to do stuff

También es común usar un rango como iterable en un <b>for</b>

for i in range(0,10):<br>
&nbsp;&nbsp;&nbsp;&nbsp;hace alguna cosa con i

#### Ejemplo. El complemento de una secuencia de ADN se obtiene reemplazando todo As con Ts, Ts con As, Gs con Cs y Cs con Gs por Gs

In [4]:
dna = "AGCAGGTGCGAGTAGCAGTAGCA"
dna

'AGCAGGTGCGAGTAGCAGTAGCA'

In [5]:
#Transformar todas las A's en Ts, Ts en As, Gs en Cs y Cs en Gs
complemento=""  #Iniciando string vacia
for x in dna:  #A cada iteración en este bucle, x será una letra del dna
    if x=="G":  #Si x fuera G, concatena la string complemento con un C
        complemento = complemento + "C"
    elif x=="C": #Si x fuera C, concatena la string complemento con un T
        complemento = complemento + "G"
    elif x == "A":
        complemento = complemento + "T"
    elif x == "T":
        complemento = complemento + "A"
        
print(complemento)

TCGTCCACGCTCATCGTCATCGT


<b>join</b> es un método de la clase string que concatena una cadena con las cadenas de una lista.<br><br>
<b>reversed</b> devuelve una lista o cadena invertida.<br><br>
Vamos a invertir una string y usar el método de join para formar una nueva cadena con el inverso complementario de la secuencia de DNA.

In [8]:
reverso_complementar = ''.join(reversed(complemento)) #conatena la string vacia '' con el complemento invertido
print(f"DNA={dna}\nREVCOMP={reverso_complementar}")

DNA=AGCAGGTGCGAGTAGCAGTAGCA
REVCOMP=TGCTACTGCTACTCGCACCTGCT


In [ ]:
for i in range(len(dna)):
    print(f"Caracter de índice {i} é um {dna[i]}")

In [ ]:
#El ciclo for anterior también se puede hacer usando la función enumerate
for i,x in enumerate(dna):
    print(f"Caracter de índice {i} é um {x}")

### Ejercicio 9:

Escriba un programa que pida un número de dados y muestre los valores de ese número de dados, al azar.

Nota. Utilizar la libreria random.
Más información en:
[1] https://stackoverflow.com/questions/3996904/generate-random-integers-between-0-and-9

<b>Salida deseada:</b><br><br>
TIRADA DE DADOS<br>
Número de dados: 0<br>
¡Imposible!<br>
<br>
TIRADA DE DADOS<br>
Número de dados:  5<br>
Dados: 3 5 1 1 3<br>

In [4]:
from random import randrange
print('TIRADA DE DADOS')
x = int(input('Número de dados:'))
print('TIRADA DE DADOS:', end="")
for i in range(x):
    print(randrange(6), end=" ")

TIRADA DE DADOS
Número de dados:5
TIRADA DE DADOS:3 1 4 2 4 

## Ejercicio 10:

Escriba un programa que pida un número de jugadores y tire un dado para cada jugador. El jugador que tire el mayor dado ganará el juego.

<b>Salida deseada:</b><br><br>
TIRADAS DE DADO<br>
Número de jugadores: 1<br>
¡Imposible!<br>
<br>
TIRADAS DE DADO<br>
Número de jugadores: 4<br>
Jugador 1: 3<br>
Jugador 2: 3<br>
Jugador 3: 5<br>
Jugador 4: 1<br>
El ganador es el Jugador 3.<br>
<br>
TIRADAS DE DADO<br>
Número de jugadores: 3<br>
Jugador 1: 4<br>
Jugador 2: 4<br>
Jugador 3: 1<br>
Los ganadores son el Jugador 1, Jugador 2.

In [30]:
from random import randrange
print('TIRADA DE DADOS')
x = int(input('Número de jugadores:'))
numeros = []
if(x>1):
    for i in range(x):
        aux = randrange(1,7)
        print('Jugador ' , i+1,':', aux)
        numeros.append(aux)
    
    aux2=0
    for i in range(x):
        if(numeros[i]==max(numeros)):
            aux2+=1

    if(aux2<2):
        print('El ganador es el ',end= '')
        for i in range(x):
            if(numeros[i]==max(numeros)):
                print('Jugador ',i+1,end= '')

    else:
        aux2=0
        print('Los ganador son el ',end= '')
        for i in range(x):
            if(numeros[i]==max(numeros)):
                if aux2 ==0:
                    aux2 = 1
                    print('Jugador ',i+1 , end="")
                else:
                    print(', Jugador ',i+1 , end="")
    print('.')
else:
    print("¡Imposible!")    



TIRADA DE DADOS
Número de jugadores:20
Jugador  1 : 2
Jugador  2 : 6
Jugador  3 : 6
Jugador  4 : 2
Jugador  5 : 5
Jugador  6 : 1
Jugador  7 : 6
Jugador  8 : 5
Jugador  9 : 5
Jugador  10 : 5
Jugador  11 : 1
Jugador  12 : 5
Jugador  13 : 1
Jugador  14 : 6
Jugador  15 : 4
Jugador  16 : 3
Jugador  17 : 5
Jugador  18 : 4
Jugador  19 : 6
Jugador  20 : 4
El ganador son el Jugador  2, Jugador  3, Jugador  7, Jugador  14, Jugador  19.


### Ejercicio 11:

Escriba un programa que pida un número de dados y tire esa cantidad de dados. El primer jugador obtiene un punto por cada dado par. El segundo jugador obtiene un punto por cada dado impar. El jugador que saque más puntos, gana.

<b>Salida deseada:</b><br><br>
PARES Y NONES<br>
Número de dados: 0<br>
¡Imposible!<br>
<br>
PARES Y NONES<br>
Número de dados: 7<br>
Dados: 4 5 2 3 2 1 6<br>
Ha ganado el jugador de los pares.<br>
<br>
PARES Y NONES<br>
Número de dados: 8<br>
Dados: 1 4 1 3 5 6 6 2<br>
Han empatado.

In [59]:
from random import randrange
print('PARES Y NONES')
x = int(input('Número de dados: '))
numeros = []
print('Dados: ' , end = '')
pares = 0
impares = 0
if(x>0):
    for i in range(x):
        aux = randrange(1,7)
        print(aux , ' ' , end = '')
        if(aux%2 == 0):
            pares+=1
        else:    
            impares+=1
    
    print('')
    if(pares> impares):
        print ('Ha ganado el jugador de los pares.')
    if(impares> pares):
        print ('Ha ganado el jugador de los impares.')
    if(impares== pares):
        print ('Ha ganado el jugador de los impares.')
else:
    print ('¡Imposible!')
    

PARES Y NONES
Número de dados: 20
Dados: 2  2  6  2  6  6  3  6  6  1  3  4  5  5  3  4  6  6  6  3  
Ha ganado el jugador de los pares.


### Ejercicio 12:

Escriba un programa que pida un número de dados y tire esa cantidad de dados. Si no salen dos dados iguales seguidos, el jugador gana. Si salen, pierde.

<b>Salida deseada:</b><br><br>
DADOS IGUALES<br>
Número de dados: 1<br>
¡Imposible!<br>
<br>
DADOS IGUALES<br>
Número de dados: 6<br>
Dados: 3 1 1 4 6 2<br>
El jugador ha perdido.<br>
<br>
DADOS IGUALES<br>
Número de dados: 4<br>
Dados: 4 2 5 3<br>
El jugador ha ganado.<br>

In [8]:
from random import randrange
print('PARES Y NONES')
x = int(input('Número de dados: '))
numeros = []
print('Dados: ' , end = '')
iguales = 0
numAnt=0
if(x>1):
    for i in range(x):
        aux = randrange(1,7)
        print(aux , ' ' , end = '')
        if(aux==numAnt):
            iguales = 1
        else:    
            numAnt = aux

    print('')
    if(iguales == 1):
        print ('El jugador ha perdido.')
    else:
        print ('El jugador ha ganado.')
else:
    print ('¡Imposible!')

PARES Y NONES
Número de dados: 7
Dados: 6  5  1  5  4  3  3  
El jugador ha perdido.


### Ejercicio 13:

Escribir un programa que pregunte al usuario una cantidad a invertir, el interés anual y el número de años, y muestre por pantalla el capital obtenido en la inversión cada año que dura la inversión.

<b>Salida deseada:</b><br><br>
¿Cantidad a invertir?  1000<br>
¿Interés porcentual anual?  10<br>
¿Años? 5<br>
Capital tras 1 años: 1100.00<br>
Capital tras 2 años: 1210.00<br>
Capital tras 3 años: 1331.00<br>
Capital tras 4 años: 1464.10<br>
Capital tras 5 años: 1610.51<br>

In [3]:
c = float(input('¿Cantidad a invertir? '))
interes = float(input('¿Interés porcentual anual? '))
a = int(input('¿Años? '))
interes = interes/100
if(a>0 and interes>0 and c>0 ):
    for i in range(a):
        c +=c*interes 
        print("Capital tras " , i+1, "años: " , "{0:.2f}".format(c))
else:
    print ('¡Imposible!')


¿Cantidad a invertir? 1
¿Interés porcentual anual? 1
¿Años? 1
Capital tras  1 años:  1.01


# Ejercicio 14:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo ("* ") como el de más abajo, de altura el número introducido.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo) 5<br>
*<br>
\* *<br>
\* * *<br>
\* * * *<br>
\* * * * * 

In [4]:
a = int(input('Introduce la altura del triángulo (entero positivo) '))
if(a>0 ):
    for i in range(a):
        print('')
        for j in range(i+1):
            print('*' , end='')
               
else:
    print ('¡Imposible!')


Introduce la altura del triángulo (entero positivo) 6

*
**
***
****
*****
******

### Ejercicio 15:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo como el de más abajo.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  9<br>

1 <br>
3 1 <br>
5 3 1 <br>
7 5 3 1 <br>
9 7 5 3 1

In [2]:
a = int(input('Introduce el tope del triángulo (entero positivo): '))

if(a>0 ):
    if(a%2==0):
        aux1=2
        aux2=aux1
        for i in range(a):
            print('')
            if(aux1>a):
                break
            for j in range(i+1):
                print(aux2 , end=' ')
                aux2-=2
            aux1+=2
            aux2=aux1
    else:
        aux1=1
        aux2=aux1
        for i in range(a):
            print('')
            if(aux1>a):
                break
            for j in range(i+1):
                print(aux2 , end=' ')
                aux2-=2
            aux1+=2
            aux2=aux1
else:
    print ('¡Imposible!')


Introduce el tope del triángulo (entero positivo): 10

2 
4 2 
6 4 2 
8 6 4 2 
10 8 6 4 2 


### Ejercicio 16:


Escriba un programa que pida un número entero mayor que cero y que escriba sus divisores.

<b>Salida deseada:</b><br><br>
DIVISORES<br>
Escriba un número mayor que cero: -5<br>
¡Le he pedido un número entero mayor que cero!<br>
<br>
DIVISORES<br>
Escriba un número entero mayor que cero: 200<br>
Los divisores de 200 son 1 2 4 5 8 10 20 25 40 50 100 200

In [4]:
a = int(input('Introduce la altura del triángulo (entero positivo): '))

if(a>0 ):
    print('Los divisores de 200 son',end=' ')
    for i in range(a):
        if(i>0 and a%i==0 ):
            print(i,end=' ')
    print(a)
else:
    print ('¡Le he pedido un número entero mayor que cero!')


Introduce la altura del triángulo (entero positivo): 521
Los divisores de 200 son 1 521


### Ejercicio 17:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo como el de número de primos.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  20<br>

1 <br>
3 1 <br>
5 3 1 <br>
7 5 3 1 <br>
11 7 5 3 1 <br>
13 11 7 5 3 1 <br>
17 13 11 7 5 3 1 <br>
19 17 13 11 7 5 3 1

In [5]:
num = int(input('Introduce la altura del triángulo (entero positivo): '))
listaPrimos=[]

for a in range( 2, num+1):
    EsPrimo=1
    for b in range( 2, a):
        if(b!=a and a%b==0):
            EsPrimo=0
    if(EsPrimo):
        listaPrimos.extend([a])
cont=0
if(num>0 ):
    for i in range(num):
        print('')
        for j in range(i+1):
            print(listaPrimos[cont-j] , end=' ')
        cont+=1
        if(listaPrimos[len(listaPrimos)-1]== listaPrimos[i]):
            break;
else:
    print ('¡Imposible!')



Introduce la altura del triángulo (entero positivo): 25

2 
3 2 
5 3 2 
7 5 3 2 
11 7 5 3 2 
13 11 7 5 3 2 
17 13 11 7 5 3 2 
19 17 13 11 7 5 3 2 
23 19 17 13 11 7 5 3 2 

### Ejercicio 18:

Escriba un programa que pida dos números enteros y escriba la suma de todos los enteros desde el primer número hasta el segundo.

<b>Salida deseada:</b><br><br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: -5<br>
¡Le he pedido un número entero positivo!<br>
<br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: 7<br>
Escriba un número entero mayor que 7: 7<br>
¡Le he pedido un número entero mayor que 7!<br>
<br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: 30<br>
Escriba un número entero mayor que 30: 32<br>
30 + 31 + 32 = 93<br>

In [8]:
print('SUMA ENTRE VALORES')
aux=0
a = int(input('Escriba un número entero positivo: '))
if(a>-1):
    b = int(input('Escriba un número entero mayor que : '))
    if(b>a):
        for i in range(a, b):
            print(i,end=' + ')
            aux+=i
        aux+=b
        print(b,' = ' , aux)
         
    else:
        print('¡Le he pedido un número entero mayor que ',a,'!')
else:
    print('¡Le he pedido un número entero positivo!')

       
    

SUMA ENTRE VALORES
Escriba un número entero positivo: 20
Escriba un número entero mayor que : 19
¡Le he pedido un número entero mayor que  20 !


### Ejercicio 19:

Escriba un programa que pida un número entero mayor que cero y calcule su factorial.

<b>Salida deseada:</b><br><br>
FACTORIAL<br>
Escriba un número entero mayor que cero: -5<br>
¡Le he pedido un número entero mayor que cero!<br>
<br>
FACTORIAL<br>
Escriba un número entero mayor que cero: 5<br>
El factorial de 5 es 120<br>

In [10]:
print('FACTORIAL')
a = int(input('Escriba un número entero mayor que cero: '))
fac=1
if(a>0):
    for i in range(1, a+1):
        fac*=i
    print('El factorial de ',a,' es ' ,fac)
else:
    print('¡Le he pedido un número entero mayor que cero!')

FACTORIAL
Escriba un número entero mayor que cero: 1
El factorial de  1  es  1


### Ejercicio 20:

Escribir un programa que pida al usuario un número entero (del 1 al 10) y genere la matriz caracol.

<b>Salida deseada:</b><br><br>

Introduce un número (entero positivo):  3<br>

1 2 3<br>
8 9 4<br>
7 6 5<br>

In [11]:
import numpy as np
import math

a = int(input('Introduce un número (entero positivo del 1 al 10): '))
if(a>1 and a<11):
    matriz = np.zeros((a, a))
    cont=1
    vueltas = math.ceil(a/2)
    for i in range(vueltas):
        for l in range(i,a -i):
            matriz[i][l] = cont
            cont+=1
        for j in range(1+i,a-i):
            matriz[j][a-i-1] = cont
            cont+=1
        for k in range(i,a-i-1):
            matriz[a-1-i][a-2-k] = cont
            cont+=1
        for p in range(i+1,a-i-1):
            matriz[a-1-p][i] = cont
            cont+=1
else:
    print('El programa solo maneja el rango del 1 al 10')
if(a>1 and a<11):
    for n in range(a):
        for m in range(a):
            print(int(matriz[n][m]) , end=' ')
        print('')

Introduce un número (entero positivo del 1 al 10): 7
1 2 3 4 5 6 7 
24 25 26 27 28 29 8 
23 40 41 42 43 30 9 
22 39 48 49 44 31 10 
21 38 47 46 45 32 11 
20 37 36 35 34 33 12 
19 18 17 16 15 14 13 


# 2.4. La Sentencia while

Una instrucción while prueba una condición inicial y comienza a ejecutar los comandos del bloque si la condición es verdadera. Mientras la condición permanezca verdadera, los comandos de bucle se ejecutan repetidamente.

Sintaxis:

mientras condición:

     comandos

#### Ejemplo de declaración while

In [ ]:
i=0
while i < 10:
    i+=1
    print(i)

In [ ]:
# Iniciando con una lista vacia
nombres = []

# Defina una string llamada nuevo_nombre con cualquier valor diferente de "salir"
nuevo_nombre = ''

El ciclo while se ejecutará mientras el usuario ingrese nombres distintos a "salir", almacenando cada nombre ingresado en una lista.

In [ ]:
while nuevo_nombre != 'salir':
    # Pida que el usuario ingrese un nombre
    nuevo_nombre = input("Ingrese un nombre o la palabra salir: ")
    # Adiciona el nombre al final de la lista de nombres
    nombres.append(nuevo_nombre)

#Extraiga la palabra salir de la lista de names ("salir")
ultima = nombres.pop()
# Imprime la lista de nombres
print(nombres)

El control de bucles break, continue y pass permite:

In [ ]:
while True:
    line = input('Input > ')
    if line == 'done':
        break
    print(line)
print('Done!')

In [ ]:
while True:
    line = input('Input > ')
    if line[0] == '#':
        continue
    if line == 'done':
        break
    print(line)
print('Done!')

Más información:<br>
[1] https://www.pythonmania.net/es/2013/04/05/control-de-bucles-break-continue-y-pass/

### Ejercicio 21:

Escriba un programa que pida dos números enteros. El programa pedirá de nuevo el segundo número mientras no sea mayor que el primero. El programa terminará escribiendo los dos números.

<b>Salida deseada:</b><br><br>
NÚMERO MAYOR<br>
Escriba un número: 6<br>
Escriba un número mayor que 6: 6<br>
6 no es mayor que 6. Inténtelo de nuevo: 1<br>
1 no es mayor que 6. Inténtelo de nuevo: 8<br>
<br>
Los números que ha escrito son 6 y 8.

In [6]:
print('NUMERO MAYOR')
num1 =input("Escriba un número: ")
print("Escriba un número mayor que",num1, end="")
num2 =input(": ")
while( num1> num2):
    print( num2," no es mayor que ", num1,". Inténtelo de nuevo",end="")
    num2 =input(": ")
print('Los números que ha escrito son ', num1,' y ',num2,'.')

NUMERO MAYOR
Escriba un número: 7
Escriba un número mayor que 7: 4
4  no es mayor que  7 . Inténtelo de nuevo: 5
5  no es mayor que  7 . Inténtelo de nuevo: 4
4  no es mayor que  7 . Inténtelo de nuevo: 9
Los números que ha escrito son  7  y  9 .


### Ejercicio 22:

Escriba un programa que pida números decimales mientras el usuario escriba número mayores que el anterior.


<b>Salida deseada:</b><br><br>
NÚMEROS MAYORES<br>
Escriba un número: 7<br>
Escriba un número mayor que 7.0: 1<br>
1.0 no es mayor que 7.0.<br>
<br>
NÚMEROS MAYORES<br>
Escriba un número: 9.3<br>
Escriba un número mayor que 9.3: 9.3<br>
9.3 no es mayor que 9.3.<br>
<br>
NÚMEROS MAYORES<br>
Escriba un número: 4.5<br>
Escriba un número mayor que 4.5: 5<br>
Escriba otro número mayor que 5.0: 7.3<br>
Escriba otro número mayor que 5.0: 2<br>
2.0 no es mayor que 4.5.

In [8]:
print('NUMEROS MAYORES')
num1 =input("Escriba un número: ")
print("Escriba un número mayor que",num1, end="")
num2 =input(": ")
while( num1< num2):
    print("Escriba otro número mayor que", num1,end="")
    num2 =input(": ")
print(num2,'no es mayor que',num1,'.')

NUMEROS MAYORES
Escriba un número: 6
Escriba un número mayor que 6: 7.6
Escriba otro número mayor que 6: 8.44
Escriba otro número mayor que 6: 5
5 no es mayor que 6 .


### Ejercicio 23:

Escriba un programa que pida un valor límite positivo y a continuación pida números hasta que la suma de los números introducidos supere el límite inicial.


<b>Salida deseada:</b><br><br>
HASTA EL LÍMITE<br>
Escriba el valor límite: -3<br>
El límite debe ser mayor que 0. Inténtelo de nuevo: 0<br>
El límite debe ser mayor que 0. Inténtelo de nuevo: 5.5<br>
Escriba un número: 7.2<br>
Ha superado el límite. La suma de los números introducidos es 7.2.<br>
<br>
HASTA EL LÍMITE<br>
Escriba el valor límite: 12<br>
Escriba un número: 5.2<br>
Escriba otro número: -1<br>
Escriba otro número: 9<br>
Ha superado el límite. La suma de los números introducidos es 13.2.

In [13]:
print('HASTA EL LIMITE')
lim=float(input('Escriba el valor límite:'))
while(lim<=0):
    print('El límite debe ser mayor que ',lim,'. Inténtelo de nuevo',end='')
    lim = float(input(': '))
num=float(input('Escriba un número:'))
while(num<=lim):
    num += float(input('Escriba otro numero: '))
print('Ha superado el límite. La suma de los números introducidos es ',num)

HASTA EL LIMITE
Escriba el valor límite:10
Escriba un número:5
Escriba otro numero: 5
Escriba otro numero: 2
Ha superado el límite. La suma de los números introducidos es  12.0


### Ejercicio 24:

Escriba un programa que calcule la descomposición en factores primos de un número.

<b>Salida deseada:</b><br><br>
DESCOMPOSICIÓN EN FACTORES PRIMOS<br>
Escriba un número entero mayor que 1: 500<br>
Descomposición en factores primos: 2 2 5 5 5<br>
<br>
DESCOMPOSICIÓN EN NÚMEROS PRIMOS<br>
Escriba un número entero mayor que 1: 521<br>
Descomposición en factores primos: 521<br>
<br>
DESCOMPOSICIÓN EN NÚMEROS PRIMOS<br>
Escriba un número entero mayor que 1: 1<br>
1 no es mayor que 1. Inténtelo de nuevo: 720<br>
Descomposición en factores primos: 2 2 2 2 3 3 5

In [31]:
print('DESCOMPOSICIÓN EN FACTORES PRIMOS')
num=int(input('Escriba un número entero mayor que 1:'))
while(num<=1):
    print(num ,'no es mayor que 1, intentelo de nuevo',end=' ')
    num=int(input(': '))
listaPrimos=[]

for a in range( 2, num+1):
    EsPrimo=1
    for b in range( 2, a):
        if(b!=a and a%b==0):
            EsPrimo=0
    if(EsPrimo):
        listaPrimos.extend([a])
pos=0
descomposicion=''
while(num!=1):
    if(num%listaPrimos[pos]==0):
        num/=listaPrimos[pos]
        descomposicion+= str(listaPrimos[pos])+' '
    else:
        pos+=1
        
print('Descomposición en factores primos: ',descomposicion)


DESCOMPOSICIÓN EN FACTORES PRIMOS
Escriba un número entero mayor que 1:500
Descomposición en factores primos:  2 2 5 5 5 


### Ejercicio 25:

Escriba un programa que muestre números al azar del 1 al 6 mientras lo pida el usuario. El programa debe mostrar la suma de los números anteriores.

<b>Salida deseada:</b><br>

Juego del Azar<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 2 - Suma: 2<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 5 - Suma: 7<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 4 - Suma: 11<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: x<br>
Programa terminado.

In [37]:
from random import randrange
salir='no'
dado1=0
while(1):
    print("Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:",end="")
    salir = input(" ")
    if(salir!=''):
        break
    dado2= randrange(1,7)
    dado1+=dado2
    print("Tirada: ",dado2," - Suma: ",dado1)
print("Programa terminado.")

Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: 
Tirada:  6  - Suma:  6
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: 
Tirada:  4  - Suma:  10
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: 
Tirada:  6  - Suma:  16
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: s
Programa terminado.


### Ejercicio 26:

Escriba un programa que pida un número del 1 al 6, un número de jugadores y tire un dado para cada jugador. El 1er jugador obtenga el mismo número será el ganador.

<b>Salida deseada:</b><br>

JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 1<br>
¡Imposible!<br>
<br>
JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 4<br>
Jugador 1: 3<br>
Jugador 2: 3<br>
Jugador 3: 5<br>
El ganador es el Jugador 3.<br>
<br>
JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 3<br>
Jugador 1: 4<br>
Jugador 2: 4<br>
Jugador 3: 1<br>
No existe ganadores.

In [43]:
from random import randrange
print("JUEGO DE LOS DADOS")
x = int(input('Número a acertar: '))
j = int(input("Número de jugadores:"))
if(j>2 and 0<x<7):
    cont=0
    dado=0
    while(j>cont and dado!=x):
        dado=randrange(1,7)
        cont+=1
        print("Jugador ",cont,": " ,dado)
    if(dado==x):
        print('El ganador es el Jugador ',cont)
    else: 
        print('No existe ganadores.')
else:
    print('!Imposible!')



JUEGO DE LOS DADOS
Número a acertar: 6
Número de jugadores:4
Jugador  1 :  5
Jugador  2 :  1
Jugador  3 :  4
Jugador  4 :  5
No existe ganadores.


### Ejercicio 27:

Escribir un programa que genere <b>una cadena de 4 caracteres aleatorios</b> que usen minúsculas y números en una variable, pregunte al usuario si desea averiguar la cadena o utilizar fuerza bruta para conseguir la contraseña correcta. En caso de usar la opción de fuerza bruta, declarar la cantidad de iteraciones utilizadas para encontrar la contraseña por fuerza bruta.

<b>Salida deseada:</b><br>

QUEBRAR UNA CONTRASEÑA<br>

El programa generó una contraseña.<br>

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  1234<br>
¡Incorrecto!

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  pass<br>
¡Contraseña correcta!<br>
<br>
<br>
QUEBRAR UNA CONTRASEÑA<br>

El programa generó una contraseña.<br>

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  cont<br>
¡Incorrecto!

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: fb<br>
Se realizó 3465 iteraciones para encontrar la contraseña correcta: abc5<br>


In [18]:
from random import randrange
print('QUEBRAR UNA CONTRASEÑA')
password=''
def comparar(lista,string):
    EsIgual=1
    for pos in range (0,4):
        if(str(lista[pos]) !=string[pos]):
            EsIgual=0
            break;
    return EsIgual

for a in range(1,5):
    iOC= randrange(0,2)
    if(iOC == 0):
        password+=chr(randrange(97,123))
    else:
        password+=str(randrange(0,10))
opcion = ''
while(1):
    opcion=input('¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?:')
    if(opcion=='fb'):
        break;
    if(opcion=='a'):
        aux=input('Introduce la contraseña: ')
        if(aux==password):
            print('!Correcto!')
            break;
        else:
            print('!Incorrecto!')
    else:
        print('No existe esa opcion')
cont=0
password2=['a','b','c','d']
for a in range(97,133):
    cont+=1
    if(comparar(password2,password)):
        break;
    if(a>122):
        password2[0]=a-123
    else:
        password2[0]=chr(a)
    for b in range(97,133):
        cont+=1
        if(comparar(password2,password)):
            break;
        if(b>122):
            password2[1]=b-123
        else:
            password2[1]=chr(b)
        for c in range(97,133):
            cont+=1
            if(comparar(password2,password)):
                break;
            if(c>122):
                password2[2]=c-123
            else:
                password2[2]=chr(c)
            for d in range(97,133):
                cont+=1
                if(comparar(password2,password)):
                    break;
                if(d>122):
                    password2[3]=d-123
                else:
                    password2[3]=chr(d)
                    
passwordQuebrado=str(password2[0])+str(password2[1])+str(password2[2])+str(password2[3])
print('Se realizó ',cont,' iteraciones para encontrar la contraseña correcta:',passwordQuebrado)
    


QUEBRAR UNA CONTRASEÑA
¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?:fb
Se realizó  1526811  iteraciones para encontrar la contraseña correcta: 53ng


### Ejercicio 28:

Escriba un programa para jugar a adivinar un número (el ordenador "piensa" el número y el usuario tiene que adivinarlo). El programa empieza pidiendo entre qué números está el número a adivinar, se "inventa" un número al azar y después el usuario va probando valores y el programa va diciendo si son demasiado grandes o pequeños.

<b>Salida deseada:</b><br>

Valor mínimo: 0<br>
Valor máximo: 100<br>
Intente adivinar un número entero entre 0 y 100.<br>
Escriba un número: 20<br>
¡Demasiado pequeño! Inténtelo de nuevo: 30<br>
¡Demasiado grande! Inténtelo de nuevo: 27<br>
¡Acertó! Le ha costado 3 intentos.

<b>Sugerencia:</b> puede perfeccionar este programa haciendo:

* que al principio el programa se asegure de que el valor máximo es superior al valor mínimo.
* que el programa detecte "trampas", por ejemplo, si al decir 25 le decimos "mayor" y al decir 26 le decimos "menor", el programa debe decirnos que estamos haciendo trampas y dejar de jugar.

In [17]:
from random import randrange
min = int(input('Valor mínimo: '))
max = int(input('Valor maximo: '))
print('Intente adivinar un número entero entre ',min,' y ',max,'.')
numI = randrange(min,max+1)
num= int(input('Escriba un número: '))
cont=0
while(1):
    cont+=1
    if(num==numI):
        print('¡Acertó! Le ha costado ',cont,' intentos.')
        break;
    if(num<numI):
        print('¡Demasiado pequeño! Inténtelo de nuevo',end='')
        num= int(input(': '))
    if(num>numI):
        print('¡Demasiado grande! Inténtelo de nuevo',end='')
        num= int(input(': '))


Valor mínimo: 4
Valor maximo: 60
Intente adivinar un número entero entre  4  y  60 .
Escriba un número: 30
¡Demasiado grande! Inténtelo de nuevo: 20
¡Demasiado pequeño! Inténtelo de nuevo: 26
¡Demasiado grande! Inténtelo de nuevo: 24
¡Demasiado grande! Inténtelo de nuevo: 22
¡Demasiado pequeño! Inténtelo de nuevo: 23
¡Acertó! Le ha costado  5  intentos.


### Ejercicio 29:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo de Fibonacci.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  7<br>

1 <br>
1 1<br>
1 1 2<br>
1 1 2 3<br>
1 1 2 3 5<br>
1 1 2 3 5 8<br>
1 1 2 3 5 8 13<br>

In [32]:
altura = int(input('Introduce la altura del triángulo (entero positivo): '))


for i in range(0,altura):
    print('')
    ant =1
    num=1
    for j in range(0,i+1):
        if(j==0 or j==1):
            print(num,end=' ')
        if(j>1):
            aux=ant
            ant=num
            num+=aux
            print(num,end=' ')

Introduce la altura del triángulo (entero positivo): 7

1 
1 1 
1 1 2 
1 1 2 3 
1 1 2 3 5 
1 1 2 3 5 8 
1 1 2 3 5 8 13 

### Ejercicio 30:

Implemente un programa que sea capaz de calcular el resultado de aplicarla formula siguiente (n i)=n!/(i!\*(n-i)!). Para ello usted debe leer un numero por teclado que debe ser mayor e igual que cero teniendo en cuenta además que n tiene que ser necesariamente mayor que i y además calcular el factorial de un número.

In [14]:
num=int(input('Introduce el valor de n: '))
i=num+1
def factorial(x):
    fac=1
    if(x>1):
        for a in range(1,x+1):
            fac*=a
        return fac
    else:
        return fac
        
        
while(1):
    if(num<0):
        num=int(input('Debe ser >= a cero, vuelve a introducir el valor de "n"'))
    else: 
        if(i>num):
            i=int(input('Introduce el valor de "i" Debe ser menor a "n": '))
        else:
            break;
resultado= factorial(num)/ (factorial(i)* factorial(num-i))
print("El resultado de la formula (n i)=n!/(i!*(n-i)!) es:", resultado)

Introduce el valor de n: 7
Introduce el valor de "i" Debe ser menor a "n": 5
El resultado de la formula (n i)=n!/(i!*(n-i)!) es: 21.0


## 